src/main/kotlin/org/rust/cargo/util/CargoCommandCompletionProvider.kt

```diff
@@ -62,9 +65,25 @@ private fun getCompleterForOption(name: String): ArgCompleter? = when (name) {
     "bench" -> targetCompleter(CargoWorkspace.TargetKind.Bench)
     "package" -> { ctx -> ctx.currentWorkspace?.packages.orEmpty().map { it.lookupElement } }
     "manifest-path" -> { ctx -> ctx.projects.map { it.lookupElement } }
+    "target" -> { ctx -> getTargetTripleCompletions(ctx) }
     else -> null
 }
 
+private fun getTargetTripleCompletions(ctx: Context): List<LookupElement> {
+    val path = ctx.currentWorkspace?.workspaceRootPath ?: return emptyList()
+    val rustup = ctx.projects.firstOrNull()?.project?.toolchain?.rustup(path) ?: return emptyList()
+
+    return rustup.listTargets().map {
```

Comment:
Note, running external process here:
- may take quite a lot of time
- can be executed under read action
```
java.lang.Throwable: Synchronous execution under ReadAction: /Users/Arseniy.Pendryak/.cargo/bin/rustup target list, see com.intellij.execution.process.OSProcessHandler#checkEdtAndReadAction() Javadoc for resolutions
	at com.intellij.openapi.diagnostic.Logger.error(Logger.java:159)
	at com.intellij.execution.process.OSProcessHandler.checkEdtAndReadAction(OSProcessHandler.java:173)
	at com.intellij.execution.process.OSProcessHandler.waitFor(OSProcessHandler.java:111)
	at com.intellij.execution.process.CapturingProcessRunner.runProcess(CapturingProcessRunner.java:122)
	at com.intellij.execution.process.CapturingProcessRunner.runProcess(CapturingProcessRunner.java:81)
	at com.intellij.execution.process.CapturingProcessHandler.runProcessWithProgressIndicator(CapturingProcessHandler.java:84)
	at org.rust.openapiext.CommandLineExtKt.runProcess(CommandLineExt.kt:115)
	at org.rust.openapiext.CommandLineExtKt.runProcessWithGlobalProgress(CommandLineExt.kt:106)
	at org.rust.openapiext.CommandLineExtKt.execute(CommandLineExt.kt:33)
	at org.rust.openapiext.CommandLineExtKt.execute$default(CommandLineExt.kt:29)
	at org.rust.cargo.toolchain.tools.Rustup.listTargets(Rustup.kt:68)
	at org.rust.cargo.util.CargoCommandCompletionProviderKt.getTargetTripleCompletions(CargoCommandCompletionProvider.kt:76)
	at org.rust.cargo.util.CargoCommandCompletionProviderKt.access$getTargetTripleCompletions(CargoCommandCompletionProvider.kt:1)
	at org.rust.cargo.util.CargoCommandCompletionProviderKt$getCompleterForOption$3.invoke(CargoCommandCompletionProvider.kt:68)
	at org.rust.cargo.util.CargoCommandCompletionProviderKt$getCompleterForOption$3.invoke(CargoCommandCompletionProvider.kt)
	at org.rust.cargo.util.RsCommandCompletionProvider.complete(RsCommandCompletionProvider.kt:70)
	at org.rust.cargo.util.RsCommandCompletionProvider.addCompletionVariants(RsCommandCompletionProvider.kt:41)
	at com.intellij.util.TextFieldCompletionProvider.fillCompletionVariants(TextFieldCompletionProvider.java:64)
	at com.intellij.util.textCompletion.TextCompletionContributor.fillCompletionVariants(TextCompletionContributor.java:58)
	at com.intellij.codeInsight.completion.CompletionService.getVariantsFromContributors(CompletionService.java:77)
	at com.intellij.codeInsight.completion.CompletionService.getVariantsFromContributors(CompletionService.java:60)
	at com.intellij.codeInsight.completion.CompletionService.performCompletion(CompletionService.java:136)
	at com.intellij.codeInsight.completion.BaseCompletionService.performCompletion(BaseCompletionService.java:41)
	at com.intellij.codeInsight.completion.CompletionProgressIndicator.lambda$calculateItems$12(CompletionProgressIndicator.java:862)
	at com.intellij.util.indexing.FileBasedIndex.lambda$ignoreDumbMode$0(FileBasedIndex.java:179)
	at com.intellij.openapi.util.RecursionManager$1.computePreventingRecursion(RecursionManager.java:111)
	at com.intellij.util.indexing.FileBasedIndexEx.ignoreDumbMode(FileBasedIndexEx.java:543)
	at com.intellij.util.indexing.FileBasedIndex.ignoreDumbMode(FileBasedIndex.java:178)
	at com.intellij.codeInsight.completion.CompletionProgressIndicator.calculateItems(CompletionProgressIndicator.java:858)
	at com.intellij.codeInsight.completion.CompletionProgressIndicator.runContributors(CompletionProgressIndicator.java:846)
	at com.intellij.codeInsight.completion.CodeCompletionHandlerBase.lambda$startContributorThread$6(CodeCompletionHandlerBase.java:353)
	at com.intellij.codeInsight.completion.AsyncCompletion.lambda$tryReadOrCancel$5(CompletionThreading.java:172)
	at com.intellij.openapi.application.impl.ApplicationImpl.tryRunReadAction(ApplicationImpl.java:1137)
	at com.intellij.codeInsight.completion.AsyncCompletion.tryReadOrCancel(CompletionThreading.java:170)
	at com.intellij.codeInsight.completion.CodeCompletionHandlerBase.lambda$startContributorThread$7(CodeCompletionHandlerBase.java:345)
	at com.intellij.codeInsight.completion.AsyncCompletion.lambda$startThread$0(CompletionThreading.java:95)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$runProcess$2(CoreProgressManager.java:178)
	at com.intellij.openapi.progress.impl.CoreProgressManager.registerIndicatorAndRun(CoreProgressManager.java:658)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:610)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:65)
	at com.intellij.openapi.progress.impl.CoreProgressManager.runProcess(CoreProgressManager.java:165)
	at com.intellij.codeInsight.completion.AsyncCompletion.lambda$startThread$1(CompletionThreading.java:91)
	at com.intellij.util.RunnableCallable.call(RunnableCallable.java:20)
	at com.intellij.util.RunnableCallable.call(RunnableCallable.java:11)
	at com.intellij.openapi.application.impl.ApplicationImpl$1.call(ApplicationImpl.java:270)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1$1.run(Executors.java:668)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1$1.run(Executors.java:665)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1.run(Executors.java:665)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

Reply:
@Undin I see at least three options how to deal with targets here:
1. Hard code the list of all possible targets. But then we will need to maintain this hard-coded list
2. Redesign `RsCommandCompletionProvider` a bit to fill completion variants asynchronously and execute `listTargets()` every time under progress
3. Execute `listTargets()` once during project update and cache the list in the `CargoProjectImpl` or something

Which way do you think is better?

Reply:
Sorry for the long response. I like variant `3`. I've recently faced the necessity to get actual toolchain name (yep, it's not the same as target but also some info from rustup) so probably we should keep some necessary info from rustup beside `rustcInfo` in `CargoProjectImpl`. Also, it may be useful for target completion in `.cargo/config`